### PR TITLE
Coding Challenge Refresh

### PR DESCRIPTION
**Pull request by nbannister-noyo**
1. update readme instructions
2. simplify models
3. remove default implementation of segments
4. new extensions.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,191 +1,297 @@
...
-## Logistics
-
-### Implementation Options
-
-You have two options for completing this exercise:
-1. **Complete it prior to your interview:** In the live session, you'll review your solution with the interviewer and work through one or more of the [Live Extensions](#live-extensions). If you choose this option we ask that you spend **no more than 1-2 hours on it**, if you are unable to complete within the time recommendation please find a stopping point and you can work through the remainder during the live interview.
-
-2. **Complete it during your interview:** We ask that you spend **20-30 minutes** familiarizing yourself with the exercise and ensuring you can start the server on your machine prior to the live interview. During the interview you will pair with the interviewer to complete the exercise and one or more of the [Live Extension](#live-extensions).
+## TLDR;
+1. You'll need docker, install it.
```
Comment (pon-noyo) [inline README.md]:
```suggestion
1. We use Docker for our local development here at Noyo, you can find the instructions for installing it on your local system [here](https://docs.docker.com/get-docker/).
```

File: README.md [hunk 2]
```diff
@@ -1,191 +1,297 @@
...
+2. Run the following... and 12 test will pass, with no warnings or errors.
+      ```sh
+      docker compose build # build some stuff
+      docker compose up -d && sleep 5 # start the server and wait
+      docker-compose exec service python seed.py # seed the db 
+      docker compose exec service pytest . # run some tests
```
Comment (pon-noyo) [inline README.md]:
```suggestion
      docker compose build # build the local containers
      docker compose up -d && sleep 5 # start up the Docker containers and wait
      docker-compose exec service python seed.py # Seed the database
      docker compose exec service pytest . # Run the test suite
```

File: README.md [hunk 3]
```diff
@@ -1,191 +1,297 @@
...
+      ```
+3. Do [Take Home Problem 1](#problem-1-implement-feature--2-hours)
```
Comment (pon-noyo) [inline README.md]:
```suggestion
3. Complete [Take Home Problem 1](#problem-1-implement-feature--1-2-hours)
```

File: README.md [hunk 4]
```diff
@@ -1,191 +1,297 @@
...
+This code was deployed to production as is but is incomplete. The original engineer was on a tight timeline and did their best; there may be bugs in the code in addition to insufficient or incorrect tests.  
 
-**Note: Please do not fork this repository or open a pull request with your submission against this repository.**
+**Note**  
+There are **no intentionally** included bugs that you have to seek out or find for this challenge. However, the incomplete tests are intentional. If you see something that you think is a bug, let's talk about it during your onsite!
 
-### Support 
 
-If you have any technical issues or questions about the instructions below please email coding-challenge@noyo.com and we will get back to you as soon as possible.
+**Part 1 (Take home - Completed before the interview) ~ 2 Hours**
```
Comment (pon-noyo) [inline README.md]:
```suggestion
**Part 1 (Take home - Completed before the interview) 1-2 Hours**  
```

File: README.md [hunk 5]
```diff
@@ -1,191 +1,297 @@
...
+      ```sh
+      docker compose build # build some stuff
+      docker compose up -d && sleep 5 # start the server and wait
+      docker-compose exec service python seed.py # seed the db 
+      docker compose exec service pytest . # run some tests
+      ```
+3. Do [Take Home Problem 1](#problem-1-implement-feature--2-hours)
+4. If you need to start over
+   ```sh
+   docker-compose down --remove-orphans --volumes # nuke all the things
```
Comment (pon-noyo) [inline README.md]:
```suggestion
   docker-compose down --remove-orphans --volumes # stop all containers from running and remove volumes
```

File: README.md [hunk 6]
```diff
@@ -1,191 +1,297 @@
...
+   ```
+5. Look at the [Live Extensions](#live-extensions), if you want, but dont do them.
```
Comment (pon-noyo) [inline README.md]:
```suggestion
5. Look at the [Live Extensions](#live-extensions), if you want, but dont do them.
5. Feel free to familiarize yourself with the [Live Extensions](#live-extensions), but please do not implement them.
```

File: README.md [hunk 7]
```diff
@@ -1,191 +1,297 @@
...
 
-### Submission Options
+# Noyo Coding Challenge! (Introduction)
 
-You have two options for completing the exercise, choose the one that you are most comfortable with:
+Within this repository is a python application that implements a fictitious product specification. Noyo is partnering with innovative insurance companies offering a new type of coverage called "traveling salesperson insurance.". This offering gives a company's employees unique benefits when they are on the road, it considers various aspects of their travel and has some very complex plan eligibility rules. When a salesman is working out of a home office, there could be gaps in their coverage timeline.
```
Comment (pon-noyo) [inline README.md:43]:
```suggestion
Within this repository is a Python application that implements a fictitious product specification. Noyo is partnering with an innovative insurance company that is offering a new type of coverage called Traveling Salesperson Insurance. This offering gives a company's employees unique benefits when they are on the road. It considers various aspects of their travel and has some complex plan eligibility rules. When a salesperson is working out of a home office, there could be gaps in their coverage timeline. 
```

File: README.md [hunk 8]
```diff
@@ -1,191 +1,297 @@
...
-### Overview
+## Problem 1. Implement Feature (~ 2 Hours)
 
-Time-based data is an important concept here at Noyo, frequently we need to understand not only what the current state of a model is, but also what it looked like at a point in the past (and sometimes the future). The purpose of this application is to store records of people and to track their address over time.
+For this exercise, you'll add functionality to the `PUT /api/persons/<person_id>/segment` endpoint, which creates a `Segment` for the `Person` with the specified ID.
 
+To begin, you will need to [start the server](#running-the-app) and create some `Person` records either by making calls to the server or running the [seed](#seeding-the-database) helper we've provided.
 
-### The Existing Codebase
+After you've created at least one `Person`, plug their `id` into the following `curl` command (or construct an API call using the tool of your choice).
```
Comment (pon-noyo) [inline README.md:66]:
```suggestion
After you've created at least one `Person`, plug their `id` into the following `curl` command (or construct an API call using the tool of your choice like [Postman](https://www.postman.com/)).
```

File: README.md [hunk 9]
```diff
@@ -1,191 +1,297 @@
...
+  -d '{
+    "start_date": "2021-01-01",
+    "city": "San Francisco",
+    "state": "CA",
+    "zip_code": "94613"
+}'
+```
 
-For this exercise, you'll add functionality to the `PUT /api/persons/<person_id>/address` endpoint, which creates an `AddressSegment` for the `Person` with the specified ID.
+_The previous API call will produce a 501 not implemented._ However its should create a single `Segment` record for the `Person` with a `start_date` of `2021-01-01` and an `end_date` of `null`. In our data model, this means that the person's location was `San Francisco, CA 94613` starting on `2021-01-01` and will remain that indefinitely. The following diagram is a visual representation that:
```
Comment (pon-noyo) [inline README.md:80]:
```suggestion
_The previous API call will produce a 501 not implemented._ However, what it should do is to create a single `Segment` record for the `Person` with a `start_date` of `2021-01-01` and an `end_date` of `null`. In our data model, this means that the person's location was `San Francisco, CA 94613` starting on `2021-01-01` and will remain that indefinitely. The following diagram is a visual representation that:
```

File: README.md [hunk 10]
```diff
@@ -1,191 +1,297 @@
...
+    "start_date": "2021-07-15",    
+    "city": "Durham",
+    "state": "NC",
+    "zip_code": "27517"
   }'
 ```
+![Figure 4](/docs/figure_4.png)
+
+
+Your challenge is to update the in segment.py to `create_segment` handling the creation of location segments.
```
Comment (pon-noyo) [inline README.md:131]:
```suggestion
Your challenge is to update `create_segment` method in the `segment.py` file to handle the creation of location segments.
```

File: README.md [hunk 11]
```diff
@@ -1,191 +1,297 @@
...
+
+### Business Rules
+1. You can only add segments to the end.
+    1. That is, adding a segment whose start date is before any current start date should result in a 422.
+2. You should treat a segment's end date of None as indefinitely in the future.
+3. A new segment can have a non-None end date.
+4. Update the end_date of an existing segment to the start_date of the new segment.
+    1. If the existing segment end_date is before the new segment's start_date, no update to the current segment is required.
+5. The endpoint should implement HTTP PUT semantics when inserting exact duplicates.
+6. If a person does not exist, return 404
```
Comment (pon-noyo) [inline README.md:141]:
```suggestion
1. You can only add segments to the end. That is, adding a segment whose start date is before any current start date should result in a `422`.
2. You should treat a segment's `end_date` of `None` as indefinitely into the future.
3. A new segment can have an `end_date` that is not `None`
4. If the `start_date` of the new segment is less than the `end_date` of the existing segment (or the `end_date` is `None`) update the the existing segment's `end_date` to match the `start_date` of the new segment.
5. The endpoint should implement [HTTP PUT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT) semantics when inserting exact duplicates.
6. If a person does not exist, return `404`
```

File: README.md [hunk 12]
```diff
@@ -1,191 +1,297 @@
...
+- Can you evaluate, review, and improve upon code that is either poor or perhaps difficult to understand and maintain.
 
-![Diagram of an address segment starting at 2021-01-01 and ending on 2021-06-15, followed by a second segment extending indefinitely](/docs/address_two_segment.png)
+*💡 There are no right or wrong answers to the concepts above, but we're interested in your thinking, arguments & results.*
 
-In order to assist with implementing this endpoint we have provided a [failing test case](/tests/api/test_addresses.py#L99) that matches the example laid out above.
+# What we are not evaluating
 
-### Live Extensions
+- Obtuse or complex python language features or skills.
```
Comment (pon-noyo) [inline README.md:157]:
```suggestion
- Obtuse or complex Python language features or skills.
```

File: README.md [hunk 13]
```diff
@@ -1,191 +1,297 @@
...
+    
+We are a python shop and expect you to have a basic understanding of python language features (or can learn them quickly), particularly those found in many other modern languages.
```
Comment (pon-noyo) [inline README.md:159]:
```suggestion
We primarily develop using Python here at Noyo, we expect you to have a basic understanding of the Python language features (or can learn them quickly), particularly those found in many other modern languages. 
```

File: README.md [hunk 14]
```diff
@@ -1,191 +1,297 @@
...
+    
+We are a python shop and expect you to have a basic understanding of python language features (or can learn them quickly), particularly those found in many other modern languages. 
 
-**Note: During your live coding interview, we will ask you to implement one or more of the following extensions to the code. You should not implement any of these extensions prior to the interview. You are welcome to look at them if you'd like, but we don't want you to spend any additional time implementing them beforehand.**
+Hint: Brush up on classes, functions, loops, if-else control flow, iterators, dictionaries, lists, tuples, datetime.date
 
-* [Extension One: `start_date` Validation](/docs/extension_one.md)
-* [Extension Two: Get Address by Date](/docs/extension_two.md)
-* [Extension Three: Merge Contiguous Identical Address Segments](/docs/extension_three.md)
+*This challenge should not require in-depth python knowledge or a deep understanding of any third-party Library. To that end, here are is a quick [Cheat sheet](/docs/cheat_sheet.md) designed to get you over hurdles that are not particularly part of this evaluation*
```
Comment (pon-noyo) [inline README.md:163]:
```suggestion
*This challenge should not require in-depth Python knowledge or a deep understanding of any third-party Library. To that end, here are is a quick [Cheat sheet](/docs/cheat_sheet.md) designed to get you over hurdles that are not particularly part of this evaluation*
```

File: README.md [hunk 15]
```diff
@@ -1,191 +1,297 @@
...
+
+- **Note: Please do not fork this repository or open a pull request with your submission against this repository.**
+
+## Support
+
+If you have any issues or questions about the instructions below, please email **coding-challenge@noyo.com**, and we will get back to you as soon as possible.
+
+## Live Extensions
+
+We will ask you to implement one or more extensions to the code during your live coding interview. You are welcome to look at them, and encouraged to do so. However, you need not spend additional time implementing them beforehand.
```
Comment (pon-noyo) [inline README.md:184]:
```suggestion
We will ask you to implement one or more extensions to the code during your live coding interview. You are welcome to look at them, and encouraged to do so. However, please do not spend additional time implementing them beforehand.
```

File: README.md [hunk 16]
```diff
@@ -1,191 +1,297 @@
...
+
 docker-compose exec service python seed.py
+
 ```
 
-*Note: You can only run the seed after you have started the server using `docker-compose up`*
+- *Note: You can only run the seed after you have started the server using `docker-compose up`*
 
-### Connecting to the Database
+### Connecting to the database
```
Comment (pon-noyo) [inline README.md:241]:
```suggestion
### Connecting to the Database
```

File: docs/cheat_sheet.md [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+
+@app.route("/api/some_route", methods=["POST"])
+@use_args(ArtistSchema())
+def some_route(payload: dict):
+    # marshmallow will validate the incomming POST payload, 
+    # ensuring that there is a valid string value.
+    return payload["name"]
+```
+
+# Flask SQL Alchemy
```
Comment (pon-noyo) [inline docs/cheat_sheet.md]:
```suggestion
# Flask SQLAlchemy
```

File: docs/cheat_sheet.md [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+We use sqlalchemy at noyo but do not expect that to have worked with it, as such here is the simple crash course on it. Please note that no sqlalchemy code is required to solve any of the problems or extenstions. This is here solely for your reference.
```
Comment (pon-noyo) [inline docs/cheat_sheet.md]:
```suggestion
We use [SQLAlchemy](https://www.sqlalchemy.org/) at Noyo, but do not expect you to have worked with it, as such here is the simple crash course on it. Please note that no advanced SQLAlchemy code is required to solve any of the problems or extensions. For example, in lieu of using the `.filter_by` syntax feel free to load models into memory and to filter using pure Python. 
```

File: docs/extension_two.md [hunk 19]
```diff
@@ -1,8 +1,7 @@
-# Extension Two: Get Address by Date
+# Extension Two: Insert and Merge address segments
```
Comment (pon-noyo) [inline docs/extension_two.md]:
```suggestion
# Extension Two: Insert and Merge Location Segments
```

File: docs/extension_two.md [hunk 20]
```diff
@@ -1,8 +1,7 @@
...
 
-You may have noticed that the `GET /api/persons/:id/address` endpoint only returns the latest address segment. In most cases this is the desired behavior, but in some cases we may want to go back in time (or forward) to see what the address was or will be. The ability to query the address by date should be implemented as follows:
+There is an external application that noyo syncs data with. They get segment data from the past that we need to inject. As such, we need a new endpoint that will take in a single segment for an employee and merge it into our existing list of segments.
```
Comment (pon-noyo) [inline docs/extension_two.md]:
```suggestion
There is an external application that Noyo syncs data with. They get segment data from the past that we need to inject. As such, we need a new endpoint that will take in a single segment for an employee and merge it into our existing list of segments. 
```

File: docs/extension_two.md [hunk 21]
```diff
@@ -1,8 +1,7 @@
-# Extension Two: Get Address by Date
+# Extension Two: Insert and Merge address segments
 
-You may have noticed that the `GET /api/persons/:id/address` endpoint only returns the latest address segment. In most cases this is the desired behavior, but in some cases we may want to go back in time (or forward) to see what the address was or will be. The ability to query the address by date should be implemented as follows:
+There is an external application that noyo syncs data with. They get segment data from the past that we need to inject. As such, we need a new endpoint that will take in a single segment for an employee and merge it into our existing list of segments. 
 
-`GET /api/persons/:id/address?date=YYYY-MM-DD`
-
-The endpoint should return the address segment that corresponds to the `date` provided in the query string. In order to assist with implementing this behavior we have provided a [failing test case](/tests/api/test_addresses.py#L140).
+1. Incoming segments all have end dates.
```
Comment (pon-noyo) [inline docs/extension_two.md]: 1. The incoming segment will always have an `end_date`

File: README.md [hunk 22]
```diff
@@ -1,191 +1,297 @@
...
-
-### Implementation Options
-
-You have two options for completing this exercise:
-1. **Complete it prior to your interview:** In the live session, you'll review your solution with the interviewer and work through one or more of the [Live Extensions](#live-extensions). If you choose this option we ask that you spend **no more than 1-2 hours on it**, if you are unable to complete within the time recommendation please find a stopping point and you can work through the remainder during the live interview.
-
-2. **Complete it during your interview:** We ask that you spend **20-30 minutes** familiarizing yourself with the exercise and ensuring you can start the server on your machine prior to the live interview. During the interview you will pair with the interviewer to complete the exercise and one or more of the [Live Extension](#live-extensions).
+## TLDR;
+1. We use Docker for our local development here at Noyo, you can find the instructions for installing it on your local system [here](https://docs.docker.com/get-docker/).
+2. Run the following... and 12 test will pass, with no warnings or errors.
```
Comment (tim-mclaughlin-noyo) [inline README.md]:
```suggestion
2. Run the following... and 12 tests will pass, with no warnings or errors.
```

File: README.md [hunk 23]
```diff
@@ -1,191 +1,297 @@
...
+```
 
-For this exercise, you'll add functionality to the `PUT /api/persons/<person_id>/address` endpoint, which creates an `AddressSegment` for the `Person` with the specified ID.
+_The previous API call will produce a 501 not implemented._ However its should create a single `Segment` record for the `Person` with a `start_date` of `2021-01-01` and an `end_date` of `null`. In our data model, this means that the person's location was `San Francisco, CA 94613` starting on `2021-01-01` and will remain that indefinitely. The following diagram is a visual representation that:
 
-To begin you will need to [start the server](#running-the-app) and create some `Person` records either by making calls to the server or running the [seed](#seeding-the-database) helper we've provided.
+![Figure 1](/docs/figure_1.png)
 
-After you've created at least one `Person`, plug their `id` into the following `curl` command (or construct an API call using the tool of your choice).  
+Making another API call that updates the person's location to `San Francisco, CA 94911` starting on `2021-04-15` and ending one `2021-08-15`. Should add this new segment to the person's existing segment list.
```
Comment (tim-mclaughlin-noyo) [inline README.md:84]:
```suggestion
Making another API call that updates the person's location to `San Francisco, CA 94911` starting on `2021-04-15` and ending one `2021-08-15`, should add this new segment to the person's existing segment list.
```

**Review by tim-mclaughlin-noyo — DISMISSED** [review]

**Review by tim-mclaughlin-noyo — APPROVED** [review]

**Pull request merged**